File: bridge/zig/src/identity.zig
const std = @import("std");
const c = @cImport({
    @cInclude("sys/utsname.h");
    @cInclude("unistd.h");
    @cInclude("pwd.h");
});

pub const VERSION = "0.1.0";

pub const Identity = struct {
    edge_version: []const u8 = VERSION,
    os: []const u8,
    arch: []const u8,
    hostname: []const u8,
    kernel: []const u8,
    user: []const u8,
    shell: []const u8,
    home: []const u8,
    cwd: []const u8,

    pub fn gather(buf: *Buffer) Identity {
        _ = c.uname(&buf.uname);

        const pw = c.getpwuid(c.getuid());
        @memset(&buf.cwd, 0);
        _ = c.getcwd(&buf.cwd, buf.cwd.len);

        return .{
            .os = span(&buf.uname.sysname),
            .arch = span(&buf.uname.machine),
            .hostname = span(&buf.uname.nodename),
            .kernel = span(&buf.uname.release),
            .user = if (pw) |p| std.mem.span(p.*.pw_name) else "unknown",
            .shell = if (pw) |p| std.mem.span(p.*.pw_shell) else "/bin/sh",
            .home = if (pw) |p| std.mem.span(p.*.pw_dir) else "/",
            .cwd = span(&buf.cwd),
        };
    }

    fn span(arr: anytype) []const u8 {
        const ptr: [*:0]const u8 = @ptrCast(arr);
        return std.mem.span(ptr);
    }

    pub fn toJson(self: Identity, buf: *Buffer) []const u8 {
        return std.fmt.bufPrint(&buf.json,
            \\{{"type":"identity","data":{{"edge_version":"{s}","os":"{s}","arch":"{s}","hostname":"{s}","kernel":"{s}","user":"{s}","shell":"{s}","home":"{s}","cwd":"{s}"}}}}
        , .{
            self.edge_version,
            self.os,
            self.arch,
            self.hostname,
            self.kernel,
            self.user,
            self.shell,
            self.home,
            self.cwd,
        }) catch "{}";
    }
};

pub const Buffer = struct {
    uname: c.utsname = std.mem.zeroes(c.utsname),
    cwd: [512]u8 = undefined,
    json: [1024]u8 = undefined,
};

File: bridge/c/identity.c
#define _POSIX_C_SOURCE 200809L
#include "identity.h"

#include <pwd.h>
#include <stdio.h>
#include <string.h>
#include <sys/utsname.h>
#include <unistd.h>

void bridge_identity_gather(bridge_identity_t *id) {
    memset(id, 0, sizeof(*id));

    struct utsname un;
    if (uname(&un) == 0) {
        snprintf(id->os,       sizeof(id->os),       "%s", un.sysname);
        snprintf(id->arch,     sizeof(id->arch),     "%s", un.machine);
        snprintf(id->hostname, sizeof(id->hostname), "%s", un.nodename);
        snprintf(id->kernel,   sizeof(id->kernel),   "%s", un.release);
    } else {
        snprintf(id->os,       sizeof(id->os),       "unknown");
        snprintf(id->arch,     sizeof(id->arch),     "unknown");
        snprintf(id->hostname, sizeof(id->hostname), "unknown");
        snprintf(id->kernel,   sizeof(id->kernel),   "unknown");
    }

    struct passwd *pw = getpwuid(getuid());
    if (pw) {
        snprintf(id->user,  sizeof(id->user),  "%s", pw->pw_name ? pw->pw_name : "unknown");
        snprintf(id->shell, sizeof(id->shell), "%s", pw->pw_shell ? pw->pw_shell : "/bin/sh");
        snprintf(id->home,  sizeof(id->home),  "%s", pw->pw_dir ? pw->pw_dir : "/");
    } else {
        snprintf(id->user,  sizeof(id->user),  "unknown");
        snprintf(id->shell, sizeof(id->shell), "/bin/sh");
        snprintf(id->home,  sizeof(id->home),  "/");
    }

    if (!getcwd(id->cwd, sizeof(id->cwd))) {
        snprintf(id->cwd, sizeof(id->cwd), "/");
    }
}

int bridge_identity_json(char *out, size_t out_cap) {
    bridge_identity_t id;
    bridge_identity_gather(&id);

    int n = snprintf(out, out_cap,
        "{\"type\":\"identity\",\"data\":{"
        "\"edge_version\":\"%s\","
        "\"os\":\"%s\","
        "\"arch\":\"%s\","
        "\"hostname\":\"%s\","
        "\"kernel\":\"%s\","
        "\"user\":\"%s\","
        "\"shell\":\"%s\","
        "\"home\":\"%s\","
        "\"cwd\":\"%s\"}}",
        BRIDGE_VERSION, id.os, id.arch, id.hostname, id.kernel,
        id.user, id.shell, id.home, id.cwd);

    if (n < 0 || (size_t)n >= out_cap) return -1;
    return n;
}
